### ISSUE TITLE
haskell-language-server missing wrapping scripts checking ABI compatibility

### ISSUE THREAD

**Issue by cdepillabout** [NixOS/nixpkgs#175565]
At some point haskell-language-server (HLS) gained a wrapper bash script that helpfully checks GHC ABI compatibility between HLS itself, and the compiler you are trying to use it with:

https://github.com/haskell/haskell-language-server/blob/6742c205c8386a2898d3a8baa830ecc75713845d/bindist/wrapper.in#L77

I believe this was introduced in HLS-1.7.0.0 (but I didn't dig through the history to confirm).

When installing HLS through `ghcup`, HLS is installed with this wrapper script.  The following is on Ubuntu with ghcup:

```console
$ which haskell-language-server-8.10.7
~/.ghcup/bin/haskell-language-server-8.10.7
$ file ~/.ghcup/bin/haskell-language-server-8.10.7
~/.ghcup/bin/haskell-language-server-8.10.7: symbolic link to haskell-language-server-8.10.7~1.7.0.0
$ file ~/.ghcup/bin/haskell-language-server-8.10.7~1.7.0.0
~/.ghcup/bin/haskell-language-server-8.10.7~1.7.0.0: symbolic link to ../hls/1.7.0.0/bin/haskell-language-server-8.10.7
$ file ~/.ghcup/hls/1.7.0.0/bin/haskell-language-server-8.10.7 
~/.ghcup/hls/1.7.0.0/bin/haskell-language-server-8.10.7: POSIX shell script, ASCII text executable, with very long lines (1639)
$ head ~/.ghcup/hls/1.7.0.0/bin/haskell-language-server-8.10.7
#!/bin/sh

exedir="/home/illabout/.ghcup/hls/1.7.0.0/lib/haskell-language-server-1.7.0.0/bin/"
executablename="haskell-language-server-8.10.7"
GHC_VERSION="8.10.7"
ABI_HASHES="Cabal-3.2.1.0:9aa4b9691119bc22fff33e788e96f702 array-0.5.4.0:84daffa90f3ca103cdf9f63b2e3dc42b ...
```

I think it would be nice to include this script in HLS from Nixpkgs, since it is so easy to accidentally pull in an HLS that has been compiled with a GHC that produces different abi hashes than you are trying to use in your own project.

cc @maralorn and @sternenseemann (since I think you guys have done a lot of the HLS work)

------------------------------------

There are also a lot of issues upstream that discuss the problems with HLS/GHC ABI-compatibility, although none are really directly related to this feature request:

- https://github.com/haskell/haskell-language-server/issues/2865
- https://github.com/haskell/haskell-language-server/issues/2903
- https://github.com/haskell/haskell-language-server/issues/2900

**Comment by maralorn** [comment]
@cdepillabout I have followed your discussion a bit. But I am not sure I understand all implications here. Do you know if this would catch mismatches between the glibc versions that hls and ghc are linked against?

**Comment by maralorn** [comment]
Anyways, yeah, this looks cool and we should probably do this. Although I wonder a bit what "this was introduced" means. Because it apparently is not part of the hackage distribution?

**Comment by cdepillabout** [comment]
@maralorn Those are good questions, but I don't really have any good answers for you.  Let me try to explain what I do know.

My company recently updated our code-base to LTS-19 / GHC-9.0.2.  One of the problems a few people ran into is that apparently the HLS-1.7.0.0 bindist built with GHC-9.0.2 and installed by ghcup is not "compatible" with the GHC-9.0.2 bindist that `stack` uses.  Aside from the above issues on the HLS issue tracker, the only place this appears to currently be documented is:

https://github.com/haskell/vscode-haskell/tree/1a7cfed578e94780cd61f59a49dda5f7cea136bd#ghc-abis-dont-match

The work-around that people are using is to use `ghcup` to compile HLS locally, instead of just using the HLS bindist:

```console
$ cabal update
$ ghcup compile hls --jobs $(nproc) --git-ref master --ghc 9.0.2
```

Although, I'm not sure why this works, since `ghcup` should still be using a ghc bindist to do this compilation of HLS, and that ghc bindist should still be incompatible with the ghc-9.0.2 bindist used by stack for LTS-19.

(We're not actually using HLS from Nixpkgs at work, but I did think that this abi-hash checker script thing could be useful for the HLS in Nixpkgs, which is why I opened this issue.)

---------------------------------------

> Do you know if this would catch mismatches between the glibc versions that hls and ghc are linked against?

My hope is that it would, although I don't know for sure.

The wrapper script introduced by HLS basically checks the abi hashes of the boot packages of the GHC it was compiled with, and compares them to the abi hashes of the boot packages of the GHC you're trying to use in your project.  I think the following line gets replaced at HLS build-time with the abi hashes of the boot packages of GHC that HLS was built with:

https://github.com/haskell/haskell-language-server/blob/6742c205c8386a2898d3a8baa830ecc75713845d/bindist/wrapper.in#L6

And then the actual calculation of your own local GHC boot package abis, along with the comparison to the HLS GHC boot package abis happens here:

https://github.com/haskell/haskell-language-server/blob/6742c205c8386a2898d3a8baa830ecc75713845d/bindist/wrapper.in#L77

As to whether or not this would catch mismatches between the glibc versions that hls and ghc are linked against, I'm not sure.  It depends on whether the system libraries that GHC is linked to are used in the calculation of the abi hash.  I'm not sure exactly what goes into the calculation of the abi hash.  Maybe someone like @sternenseemann knows this?

It would be unfortunate if the above abi-hash checks don't catch things like different glibc versions.  But, even if they don't, the abi-hash checks will still likely catch things like:

- Using the same version of GHC, but compiled with different options (like profiling vs. no profiling)
- Using the same version of GHC, but where one GHC is `haskell.compiler.ghc902` and the other GHC is `haskell.compiler.native-bignum.ghc902`.  (This can happen if a user accidentally installs HLS globally, but then tries to use `pkgsStatic.haskellPackages` in one of their projects.)

> Although I wonder a bit what "this was introduced" means. Because it apparently is not part of the hackage distribution?

I'm also not sure about this.  I know that if you install HLS with `ghcup` you get the wrapper script.  But I'm not sure if `ghcup` is doing something special here or what.  I imagine we could open an issue on the HLS repo asking for this script to be included in the hackage sdist if it is not already.

**Comment by maralorn** [comment]
@cdepillabout Thank you for the explanation.

This seems like an improvement either way. I just feel like it would take (at least me) a while to figure out how to do it.

The ABI_HASHES seem to get generated somewhere in here: https://github.com/haskell/haskell-language-server/blob/b5a37f7fc360596899cb2945f363030f44156415/GNUmakefile

**Comment by maralorn** [comment]
Heads up: I have tried to implement this in nixpkgs. See: https://github.com/maralorn/nixpkgs/commit/e8d3ab3625a79c76cd27495adae2312e4cee6f71

I think my change is theoretically doing the right thing. Practically, it’s not working at all.

The issue here is, that the abi-script basically compares the output of `ghc-pkg list` during buildtime of hls with the output at runtime. But we in the nixverse use the ghc-pkg database extensively for customization. Every interactive nix-shell or every ghcWithPackages will have a different list of packages installed and they will all be listed by ghc-pkg. (I didn‘t fully understand this before, but now I think I mainly got it.)

A side effect of this is, that e.g. in my ghcWithPackages environment I have so many packages in the database that running the wrapper errors only after 30 seconds. (something something bash not lazy …)

This means that the approach used by ghcup won‘t cut it for us. The next best idea I have would be to check if the packages present at install time are a subset of the packages present at runtime time. I am not certain if that would be a smart idea. @hasufell do you have advise to give on this?

Anyways if we decide to go that route we probably need to copy the wrapper script into nixpkgs because at that point we are probably rewriting quite a bit of it. Or we upstream the changes if we manage to make them general enough, this is in theory an issue which could happen outside of the nixverse, too …

Here is an example error output of the script, btw.:

```

❯ HLS_WRAPPER_DEBUG=1 ~/git/haskell-updates/result/bin/haskell-language-server
Trying method: ghc-9.0.2 in PATH
GHC ABIs don't match!

Expected: Cabal-3.4.1.0:e59e5d6b8995fdab4598255fd8c97130 array-0.5.4.0:993d119e4809f52f78cf2592167d72e4 base-4.15.1.0:402510cb50fcb6bd27837cb25d951868 binary-0.8.8.0:509c2c62b0668a038fd6686e64b3f670 bytestring-0.10.12.1:f5a2cac57d541fec04564d3687111a85 containers-0.6.4.1:5f3358d5f710da50369a9ce99cd36b87 deepseq-1.4.5.0:1210a92194b0c15678ea293f1b399cdb directory-1.3.6.2:59ee307d60dd0c645625b8af0833d897 exceptions-0.10.4:c5cb5ce761f458e46bdd9f40eaefa8e7 filepath-1.4.2.1:85d9509cf60b0e5e0bc78e7a1a13ea4c ghc-9.0.2:9642114b30aec33eb7a3ccfa17977d25 ghc-bignum-1.1:a1196aa0b5995b9e74349e74473e14cd ghc-boot-9.0.2:af1792a0246154b0e91ca9d1bd2a6915 ghc-boot-th-9.0.2:a065e81aed35bdef7d6a5c089521c514 ghc-compact-0.1.0.0:27402a6bc9f063ac8066cb3d4131ee33 ghc-heap-9.0.2:d0181bd91eb26c48738bfca25034b558 ghc-prim-0.7.0:37b8b91d0e1740da9fc9bd47e9677ad3 ghci-9.0.2:faa784ee8d2f8598e60003171961baea haskeline-0.8.2:609366aaf33fa47dd6e95975769a7fb4 hpc-0.6.1.0:e31358248ac35866b4dcf8edcd645ecb integer-gmp-1.1:71a531b6bf4691fcc47fea9fcdcd7b90 libiserv-9.0.2:5898d1ba2bf7e88e26541b5eb83cf051 mtl-2.2.2:4e4073b3a833beafa06fdaf350f748e1 parsec-3.1.14.0:1d1671bfcf4709bd9e39facbcbd07494 pretty-1.1.3.6:0abbc7e400e2de7bc051bb2e7b67b5df process-1.6.13.2:111749f087f24c8b7388f3db86a0247a rts-1.0.2: stm-2.5.0.0:8504c8dbccbde5ceb1d128c4b0ed878a template-haskell-2.17.0.0:cafc7e1f2f072741c31fbeed1f32e47e terminfo-0.4.1.5:bb3516a0af5e158051695709b2d32a7c text-1.2.5.0:c66d668dea4ecfed630f4bac74b3792d time-1.9.3:4584928d51136b00b8816ee4792d1f30 transformers-0.5.6.2:756632d1a506a0d16c4ea26dc4c58bda unix-2.7.2.2:dbbad7e7e153f9ac1fa6ede0c09603c6 xhtml-3000.2.2.1:7db7bc1ba300a9568cd8c17d0b6114b1
Got:      Cabal-3.4.1.0:e59e5d6b8995fdab4598255fd8c97130 Cabal-3.8.1.0:ade0c1e541e6533f9a12e737581a36d2 Cabal-syntax-3.6.0.0:b35f82855b602f4b74515d074a43bd75 Cabal-syntax-3.8.1.0:1fc2acff28843b7784e9638c1b9f5e8d Diff-0.4.1:61d97a8e18d0196976d3a37f89ad221e Glob-0.10.2:8219bf89b1a4f474f00d1eac733c453d HTTP-4000.3.16:d00f4e7003189547d039beeac64af4d5 HUnit-1.6.2.0:d6975511ec02214935e474f448c59b95 HsYAML-0.2.1.1:4d83d72bb547360508237a0b88e11081 HsYAML-aeson-0.2.0.1:8b8ab78b262d0fb21a26252adff4a4af JuicyPixels-3.3.7:134b2fe55719217819e99117e141633f ListLike-4.7.7:79aef6f8171e308bd2071a89855b4c0b MemoTrie-0.6.10:67dcd3bbc6c88484c6b03ece30059b7b MonadRandom-0.5.3:6854d5f373cf0b1043e8e740d9719793 OneTuple-0.3.1:5ba9f9f529d896054579f92e11d676f7 Only-0.1:64cf2da6a2446a1e3586cac3eb51fe49 QuickCheck-2.14.2:3c9261b25487b10957c1452a5cf03e15 RSA-2.4.1:590d0ae44c0e4b965e29239b6f987c61 SHA-1.6.4.4:7486564c89132c45bf8270dc00d36e10 StateVar-1.2.2:20f18a5af04e629a8460f999f719c270 abstract-deque-0.3:121ec24891b968a002c8002844ce3a7e abstract-par-0.3.3:8b6c04a09e47c296d23af11a0b453452 adjunctions-4.4.2:5089fa02daf5ad769fbad5653fe3d275 aeson-1.5.6.0:57b8115856efc33592f37971c8f1ae98 aeson-2.0.3.0:e2e788de99f1544a19dca9d8b477210a aeson-gadt-th-0.2.5.1:4983f044fcadbd3c591d9b3e4fa3ac6f aeson-pretty-0.8.9:49b42a03cf3e35fcc5e27d0b25999fc5 algebraic-graphs-0.6.1:7ee0ecd5fe907f762473f148bdf7c8f6 annotated-wl-pprint-0.7.0:7dbf2c0d524418e56397a5751550e08e ansi-terminal-0.11.3:0530822acacbe2cdcdbf409be13ca6f0 ansi-wl-pprint-0.6.9:0578c7acf44d5dc7c21129f8eff068be appar-0.1.8:f499b0a7c6f94be44a2336bd53c09c09 apply-refact-0.9.3.0:aa11e1902b7bffdfdff7b463c08d630e array-0.5.4.0:993d119e4809f52f78cf2592167d72e4 asn1-encoding-0.9.6:8804344c2e19c2dc9b07320f74ad5abc asn1-parse-0.9.5:7d1172708b2974b90de6380171b71963 asn1-types-0.3.4:551553745c5f28d5daa957cf60c60f21 assoc-1.0.2:4c1c7a1cad11bb04e56ffdc8314cfb91 async-2.2.4:9cdc2b62f5c2d3f427e95742ae9e6867 atomic-primops-0.8.4:1f308b1860d50182e10923f85ff7cb3c atomic-write-0.2.0.7:babcc3685b118c6f139c1f8488e468b0 attoparsec-0.14.4:44d84b68add0a185f07fa38491049efb z-attoparsec-z-attoparsec-internal-0.14.4:a394cb9ef453c8bb6848f44f8229fdc1 attoparsec-iso8601-1.0.2.1:50eca4060f35a7737d0a7ce5a14384d5 authenticate-oauth-1.7:093cd6026554feeee164db3dc41dc05c auto-update-0.1.6:6b213d592034e3eb4a18286b31057b7d base-4.15.1.0:402510cb50fcb6bd27837cb25d951868 base-compat-0.11.2:416aedce9f6e7337200a15a8123b55b0 base-compat-batteries-0.11.2:82b4107d4e91698b2c64cd03d891620f base-noprelude-4.13.0.0:b35f82855b602f4b74515d074a43bd75 base-orphans-0.8.7:45afa955a57a04253a37e1fe47211fea base16-bytestring-1.0.2.0:0ac937ca487472d9b7477848b50e3d1a 4eb54144c58374f51b27d4962f394a1c base16-bytestring-1.0.2.0:0ac937ca487472d9b7477848b50e3d1a 4eb54144c58374f51b27d4962f394a1c base64-0.4.2.4:bf16d2c0342a7e89bc330e80fcdb4f98 base64-bytestring-1.2.1.0:0bb5fbc892f95b62ecca64ee2450e829 …

**Comment by hasufell** [comment]
> The issue here is, that the abi-script basically compares the output of ghc-pkg list during buildtime of hls with the output at runtime. But we in the nixverse use the ghc-pkg database extensively for customization. Every interactive nix-shell or every ghcWithPackages will have a different list of packages installed and they will all be listed by ghc-pkg. (I didn‘t fully understand this before, but now I think I mainly got it.)

As a workaround, just compare the abi hashes of the `ghc` and `base` boot packages. That should already catch most, if not all.

**Comment by sternenseemann** [comment]
@maralorn I presume this can be closed?

**Comment by maralorn** [comment]
Correct, our wrapper script is working fine (with one minor, but fixed, hickup) for months.
